### PR TITLE
pkg/cloud/azure: migrate to new azure sdk

### PR DESCRIPTION
**Pull request by msbutler**
This patch replaces the deprecated azure sdk with azure's new sdk, used to read
and write backups to azure. This PR also reverts the custom azure put uploader, 
which if necessary, can be re added to use the new sdk in the future. 

This PR should unblock azure kms work. 

Informs #86903

Epic CRDB-18954

### CODE REVIEW

**Comment by msbutler** [comment]
cloud unit tests on this branch passed https://teamcity.cockroachdb.com/viewType.html?buildTypeId=Cockroach_Nightlies_CloudUnitTests&branch_Cockroach_Nightlies=95407&tab=buildTypeStatusDiv

**Comment by msbutler** [comment]
CI flake. will rerun after review.

File: build/bazelutil/distdir_files.bzl [hunk 1]
```diff
@@ -16,12 +16,13 @@ DISTDIR_FILES = {
...
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-sdk-for-go/sdk/azcore/com_github_azure_azure_sdk_for_go_sdk_azcore-v1.3.0.zip": "cf80995c85451a7990c4d68dfbfd7de89536d319df9502ba9dfd38eb84501810",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-sdk-for-go/sdk/azidentity/com_github_azure_azure_sdk_for_go_sdk_azidentity-v1.1.0.zip": "27947f13cb64475fd59e5d9f8b9c042b3d1e8603f49c54fc42820001c33d5f78",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-sdk-for-go/sdk/internal/com_github_azure_azure_sdk_for_go_sdk_internal-v1.1.1.zip": "10f2a543b9e000a988722c8210d30d377c2306b042e5de1bfea4b3ec730d0319",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-sdk-for-go/sdk/storage/azblob/com_github_azure_azure_sdk_for_go_sdk_storage_azblob-v0.6.1.zip": "c2539d189b22bdb6eb67c4682ded4e070d6cf0f52c8bd6899f7eb1408045783f",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-storage-blob-go/com_github_azure_azure_storage_blob_go-v0.8.0.zip": "3b02b720c25bbb6cdaf77f45a29a21e374e087081dedfeac2700aed6147b4b35",
```
Comment (benbardin) [inline build/bazelutil/distdir_files.bzl:25]: Same, this looks like a downgrade?

File: build/bazelutil/distdir_files.bzl [hunk 2]
```diff
@@ -16,12 +16,13 @@ DISTDIR_FILES = {
     "https://storage.googleapis.com/cockroach-godeps/gomod/contrib.go.opencensus.io/exporter/prometheus/io_opencensus_go_contrib_exporter_prometheus-v0.4.0.zip": "e5dc381a98aad09e887f5232b00147308ff806e9189fbf901736ccded75a3357",
     "https://storage.googleapis.com/cockroach-godeps/gomod/dmitri.shuralyov.com/gpu/mtl/com_shuralyov_dmitri_gpu_mtl-v0.0.0-20201218220906-28db891af037.zip": "d9992a60371571861985f191ac38954db2cc1bd36e834058ad50f989ccc8fa06",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/AndreasBriese/bbloom/com_github_andreasbriese_bbloom-v0.0.0-20190825152654-46b345b51c96.zip": "8b8c016e041592d4ca8cbd2a8c68e0dd0ba1b7a8f96fab7422c8e373b1835a2d",
-    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-pipeline-go/com_github_azure_azure_pipeline_go-v0.2.3.zip": "5881763a04c7705d2f348bfe18ec56266499265487d99fe448ded9874f4f3672",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-pipeline-go/com_github_azure_azure_pipeline_go-v0.2.1.zip": "83822bc4aca977af31cdb1c46012e64c819d2b9ed53885dd0f8dca5566993a5f",
```
Comment (benbardin) [inline build/bazelutil/distdir_files.bzl:19]: This looks like a downgrade?

File: build/bazelutil/distdir_files.bzl [hunk 3]
```diff
@@ -596,7 +598,7 @@ DISTDIR_FILES = {
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/matryer/is/com_github_matryer_is-v1.4.0.zip": "8c49393743a0b5c61b130632e402efa55d4b03456f31aea14a269d47503df112",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/matryer/moq/com_github_matryer_moq-v0.0.0-20190312154309-6cfb0558e1bd.zip": "b9fb2bc3d0894dfaa3cc4298f49c97346ccb66f2f0e6911f4f224ffc9acc3972",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/mattn/go-colorable/com_github_mattn_go_colorable-v0.1.12.zip": "c4637499409a4081fbbd7ea0640aa0af30c0e06a70aa1b207780b0251650d8cd",
-    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/mattn/go-ieproxy/com_github_mattn_go_ieproxy-v0.0.1.zip": "2a1ad026c677bddef4136c76da19db2204d568b4c2246b2e00a79b47d8c01192",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/mattn/go-ieproxy/com_github_mattn_go_ieproxy-v0.0.0-20190610004146-91bb50d98149.zip": "2982ad9362d63b30a081fe7609b595fefcc7baaaeda2f5f17a7dbeb087a84020",
```
Comment (benbardin) [inline build/bazelutil/distdir_files.bzl:601]: Also a downgrade? And to a custom build? (Also, is this related to azure? something about internet explorer?)

File: go.mod [hunk 4]
```diff
@@ -314,7 +315,6 @@ require (
 	github.com/lufia/plan9stats v0.0.0-20211012122336-39d0f177ccd0 // indirect
 	github.com/magiconair/properties v1.8.5 // indirect
 	github.com/mailru/easyjson v0.7.6 // indirect
-	github.com/mattn/go-ieproxy v0.0.1 // indirect
```
Comment (benbardin) [inline go.mod:317]: Oh, can this be removed from the go.sum then? And distdir files?

File: build/bazelutil/distdir_files.bzl [hunk 5]
```diff
@@ -16,12 +16,13 @@ DISTDIR_FILES = {
...
     "https://storage.googleapis.com/cockroach-godeps/gomod/dmitri.shuralyov.com/gpu/mtl/com_shuralyov_dmitri_gpu_mtl-v0.0.0-20201218220906-28db891af037.zip": "d9992a60371571861985f191ac38954db2cc1bd36e834058ad50f989ccc8fa06",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/AndreasBriese/bbloom/com_github_andreasbriese_bbloom-v0.0.0-20190825152654-46b345b51c96.zip": "8b8c016e041592d4ca8cbd2a8c68e0dd0ba1b7a8f96fab7422c8e373b1835a2d",
-    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-pipeline-go/com_github_azure_azure_pipeline_go-v0.2.3.zip": "5881763a04c7705d2f348bfe18ec56266499265487d99fe448ded9874f4f3672",
+    "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-pipeline-go/com_github_azure_azure_pipeline_go-v0.2.1.zip": "83822bc4aca977af31cdb1c46012e64c819d2b9ed53885dd0f8dca5566993a5f",
     "https://storage.googleapis.com/cockroach-godeps/gomod/github.com/Azure/azure-sdk-for-go/com_github_azure_azure_sdk_for_go-v57.1.0+incompatible.zip": "7232ccaf96ab411dce27e8c61f4ec0e20835f60e192ffc7b56b7c3a308e29978",
```
Comment (benbardin) [inline build/bazelutil/distdir_files.bzl:20]: Does this need upgrading? I think the KMS library under this version still uses the old API, for instance.

File: go.mod [hunk 6]
```diff
@@ -40,9 +40,7 @@ require (
 	cloud.google.com/go/kms v1.1.0
 	cloud.google.com/go/pubsub v1.16.0
 	cloud.google.com/go/storage v1.21.0
-	github.com/Azure/azure-pipeline-go v0.2.3 // indirect
```
Comment (benbardin) [inline go.mod:43]: Can this be removed entirely from our dependencies?

Comment (msbutler) [reply go.mod:43]: Yup. All the changes to go.mod, go.sum, and distdir_files.bzl trickle out of following the [directions ](https://github.com/cockroachdb/cockroach/tree/master/build) of updating dependencies. The only actual go code I modified which caused all of this is modifying the imports in `azure_storage.go`.

File: go.mod (hunk 4 above)
Comment (msbutler) [reply go.mod:317]: explained above.

File: build/bazelutil/distdir_files.bzl (hunk 1 above)
Comment (msbutler) [reply build/bazelutil/distdir_files.bzl:25]: it's not me, it's azure!

File: build/bazelutil/distdir_files.bzl (hunk 2 above)
Comment (msbutler) [reply build/bazelutil/distdir_files.bzl:19]: it's not me, it's azure!

File: build/bazelutil/distdir_files.bzl (hunk 3 above)
Comment (msbutler) [reply build/bazelutil/distdir_files.bzl:601]: yup it's a package azure uses. I guess the new sdk uses an older version of this. it indeed looks like an internet explorer package lol.

**Review by benbardin — APPROVED** [review]
LGTM, thanks!

**Review by dt — APPROVED** [review]

**Comment by msbutler** [comment]
bors r=benbardin, dt
